File: dqops/src/main/java/com/dqops/rest/models/comparison/TableComparisonModel.java
/*
 * Copyright © 2021 DQOps (support@dqops.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dqops.rest.models.comparison;

import com.dqops.checks.AbstractRootChecksContainerSpec;
import com.dqops.checks.CheckTimeScale;
import com.dqops.checks.CheckType;
import com.dqops.checks.comparison.*;
import com.dqops.core.jobqueue.jobs.data.DeleteStoredDataQueueJobParameters;
import com.dqops.metadata.comparisons.TableComparisonConfigurationSpec;
import com.dqops.metadata.comparisons.TableComparisonGroupingColumnsPairSpec;
import com.dqops.metadata.comparisons.TableComparisonGroupingColumnsPairsListSpec;
import com.dqops.metadata.id.HierarchyId;
import com.dqops.metadata.search.CheckSearchFilters;
import com.dqops.metadata.sources.ColumnSpec;
import com.dqops.metadata.sources.PhysicalTableName;
import com.dqops.metadata.sources.TableSpec;
import com.dqops.utils.docs.SampleStringsRegistry;
import com.dqops.utils.docs.SampleValueFactory;
import com.dqops.utils.exceptions.DqoRuntimeException;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import io.swagger.annotations.ApiModel;
import lombok.Data;

import java.util.ArrayList;
import java.util.List;

/**
 * Model that contains the all editable information about a table-to-table comparison defined on a compared table.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@ApiModel(value = "TableComparisonModel", description = "Model that contains the all editable information about a table-to-table comparison defined on a compared table.")
@Data
public class TableComparisonModel {
    /**
     * The name of the table comparison configuration that is defined in the 'table_comparisons' node on the table specification.
     */
    @JsonPropertyDescription("The name of the table comparison configuration that is defined in the 'table_comparisons' node on the table specification.")
    private String tableComparisonConfigurationName;

    /**
     * Compared connection name - the connection name to the data source that is compared (verified).
     */
    @JsonPropertyDescription("Compared connection name - the connection name to the data source that is compared (verified).")
    private String comparedConnection;

    /**
     * The schema and table name of the compared table that is verified.
     */
    @JsonPropertyDescription("The schema and table name of the compared table that is verified.")
    private PhysicalTableName comparedTable;

    /**
     * Reference connection name - the connection name to the data source that has the reference data to compare to.
     */
    @JsonPropertyDescription("Reference connection name - the connection name to the data source that has the reference data to compare to.")
    private String referenceConnection;

    /**
     * The schema and table name of the reference table that has the expected data.
     */
    @JsonPropertyDescription("The schema and table name of the reference table that has the expected data.")
    private PhysicalTableName referenceTable;

    /**
     * List of column pairs from both the compared table and the reference table that are used in a GROUP BY clause.
     */
    @JsonPropertyDescription("List of column pairs from both the compared table and the reference table that are used in a GROUP BY clause  " +
            "for grouping both the compared table and the reference table (the source of truth). " +
            "The columns are used in the next of the table comparison to join the results of data groups (row counts, sums of columns) between the compared table and the reference table to compare the differences.")
    private List<TableComparisonGroupingColumnPairModel> groupingColumns = new ArrayList<>();

    /**
     * The template of the compare thresholds that should be applied to all comparisons when the comparison is enabled.
     */
    @JsonPropertyDescription("The template of the compare thresholds that should be applied to all comparisons when the comparison is enabled.")
    private CompareThresholdsModel defaultCompareThresholds = CompareThresholdsModel.createDefaultCompareThreshold();

    /**
     * The row count comparison configuration.
     */
    @JsonPropertyDescription("The row count comparison configuration.")
    private CompareThresholdsModel compareRowCount;

    /**
     * The column count comparison configuration.
     */
    @JsonPropertyDescription("The column count comparison configuration.")
    private CompareThresholdsModel compareColumnCount;

    /**
     * Boolean flag that decides if this comparison type supports comparing the column count between tables. Partitioned table comparisons do not support comparing the column counts.
     */
    @JsonPropertyDescription("Boolean flag that decides if this comparison type supports comparing the column count between tables. Partitioned table comparisons do not support comparing the column counts.")
    private boolean supportsCompareColumnCount;

    /**
     * The list of compared columns, their matching reference column and the enabled comparisons.
     */
    @JsonPropertyDescription("The list of compared columns, their matching reference column and the enabled comparisons.")
    private List<ColumnComparisonModel> columns = new ArrayList<>();

    /**
     * Configured parameters for the "check run" job that should be pushed to the job queue in order to run the table comparison checks for this table, using checks selected in this model
     */
    @JsonPropertyDescription("Configured parameters for the \"check run\" job that should be pushed to the job queue in order to run the table comparison checks for this table, using checks selected in this model.")
    private CheckSearchFilters compareTableRunChecksJobTemplate;

    /**
     * Configured parameters for the "data clean" job that after being supplied with a time range should be pushed to the job queue in order to remove stored check results for this table comparison.
     */
    @JsonPropertyDescription("Configured parameters for the \"data clean\" job that after being supplied with a time range should be pushed to the job queue in order to remove stored check results for this table comparison.")
    private DeleteStoredDataQueueJobParameters compareTableCleanDataJobTemplate;

    /**
     * Boolean flag that decides if the current user can update or delete the table comparison.
     */
    @JsonPropertyDescription("Boolean flag that decides if the current user can update or delete the table comparison.")
    private boolean canEdit;

    /**
     * Boolean flag that decides if the current user can run comparison checks.
     */
    @JsonPropertyDescription("Boolean flag that decides if the current user can run comparison checks.")
    private boolean canRunCompareChecks;

    /**
     * Boolean flag that decides if the current user can delete data (results).
     */
    @JsonPropertyDescription("Boolean flag that decides if the current user can delete data (results).")
    private boolean canDeleteData;

    /**
     * Creates a table comparison model, copying the configuration of all comparison checks on the table level and on the column level.
     * @param comparedTableSpec Source table specification (the compared table).
     * @param referenceTableSpec Reference table specification.
     * @param tableComparisonConfigurationName The table comparison name (the reference table configuration name).
     * @param checkType Check type (profiling, monitoring, partitioned).
     * @param checkTimeScale Check time scale for monitoring and partitioned checks.
     * @param canCompareTables True if the user can edit, compare tables, delete data.
     * @return Table comparison mode.
     */
    public static TableComparisonModel fromTableSpec(TableSpec comparedTableSpec,
                                                     TableSpec referenceTableSpec,
                                                     String tableComparisonConfigurationName,
                                                     CheckType checkType,
                                                     CheckTimeScale checkTimeScale,
                                                     boolean canCompareTables) {
        TableComparisonModel tableComparisonModel = new TableComparisonModel();
        HierarchyId comparedTableHierarchyId = comparedTableSpec.getHierarchyId();
        if (comparedTableHierarchyId == null) {
            throw new DqoRuntimeException("Cannot map a detached table, because the connection and table name is unknown");
        }

        TableComparisonConfigurationSpec comparisonSpec = comparedTableSpec.getTableComparisons().get(tableComparisonConfigurationName);
        if (comparisonSpec == null) {
            throw new DqoRuntimeException("Table comparison '" + tableComparisonConfigurationName + "' not configured on the table " + comparedTableHierarchyId.toString());
        }

        tableComparisonModel.setTableComparisonConfigurationName(comparisonSpec.getComparisonName());
        tableComparisonModel.comparedConnection = comparedTableHierarchyId.getConnectionName();
        tableComparisonModel.comparedTable = comparedTableHierarchyId.getPhysicalTableName();
        tableComparisonModel.referenceConnection = comparisonSpec.getReferenceTableConnectionName();
        tableComparisonModel.referenceTable = new PhysicalTableName(comparisonSpec.getReferenceTableSchemaName(), comparisonSpec.getReferenceTableName());
        tableComparisonModel.setCanEdit(canCompareTables);
        tableComparisonModel.setCanRunCompareChecks(canCompareTables);
        tableComparisonModel.setCanDeleteData(canCompareTables);

        for (TableComparisonGroupingColumnsPairSpec groupingColumnsPairSpec : comparisonSpec.getGroupingColumns()) {
            TableComparisonGroupingColumnPairModel tableComparisonGroupingColumnPairModel =
                    TableComparisonGroupingColumnPairModel.fromColumnPairSpec(groupingColumnsPairSpec);
            tableComparisonModel.getGroupingColumns().add(tableComparisonGroupingColumnPairModel);
        }

        AbstractRootChecksContainerSpec tableCheckRootContainer = comparedTableSpec.getTableCheckRootContainer(
                checkType, checkTimeScale, false);
        AbstractComparisonCheckCategorySpecMap<?> comparisons = tableCheckRootContainer.getComparisons();
        tableComparisonModel.supportsCompareColumnCount = tableCheckRootContainer.getCheckType() != CheckType.partitioned;

        if (comparisons instanceof AbstractTableComparisonCheckCategorySpecMap) {
            AbstractTableComparisonCheckCategorySpecMap<? extends AbstractTableComparisonCheckCategorySpec> tableCheckComparisonsMap =
                    (AbstractTableComparisonCheckCategorySpecMap<? extends AbstractTableComparisonCheckCategorySpec>)comparisons;
            AbstractTableComparisonCheckCategorySpec tableCheckComparisonChecks = tableCheckComparisonsMap.get(tableComparisonConfigurationName);

            if (tableCheckComparisonChecks != null) {
                ComparisonCheckRules rowCountMatch = tableCheckComparisonChecks.getCheckSpec(TableCompareCheckType.row_count_match, false);
                tableComparisonModel.setCompareRowCount(CompareThresholdsModel.fromComparisonCheckSpec(rowCountMatch));

                if (tableCheckComparisonChecks.supportsColumnComparisonCheck()) {
                    ComparisonCheckRules columnCountMatch = tableCheckComparisonChecks.getCheckSpec(TableCompareCheckType.column_count_match, false);
                    tableComparisonModel.setCompareColumnCount(CompareThresholdsModel.fromComparisonCheckSpec(columnCountMatch));
                }
            }
        }

        for (ColumnSpec columnSpec : comparedTableSpec.getColumns().values()) {
            ColumnComparisonModel columnComparisonModel = ColumnComparisonModel.fromColumnSpec(
                    columnSpec, tableComparisonConfigurationName, checkType, checkTimeScale);
            if (columnComparisonModel.getReferenceColumnName() == null) {
                // reference column is not configured, let's try to find a column that will match
                if (referenceTableSpec != null) {
                    ColumnSpec mostSimilarReferenceColumn = referenceTableSpec.getColumns().findMostSimilarColumn(columnSpec.getColumnName());
                    if (mostSimilarReferenceColumn != null) {
                        columnComparisonModel.setReferenceColumnName(mostSimilarReferenceColumn.getColumnName());
                    }
                }
            }
            tableComparisonModel.getColumns().add(columnComparisonModel);
        }

        tableComparisonModel.compareTableRunChecksJobTemplate = new CheckSearchFilters() {{
           setConnection(comparedTableHierarchyId.getConnectionName());
           setFullTableName(comparedTableSpec.getPhysicalTableName().toTableSearchFilter());
           setCheckType(checkType);
           setTimeScale(checkTimeScale);
           setCheckCategory(AbstractComparisonCheckCategorySpecMap.COMPARISONS_CATEGORY_NAME);
           setTableComparisonName(tableComparisonConfigurationName);
           setEnabled(true);
        }};

        tableComparisonModel.compareTableCleanDataJobTemplate =
                DeleteStoredDataQueueJobParameters.fromCheckSearchFilters(tableComparisonModel.compareTableRunChecksJobTemplate);

        return tableComparisonModel;
    }

    /**
     * Copies the configuration of table comparison on the table level and columns level to correct checks, creating these checks when comparison checks are not enabled.
     * @param targetTableSpec Target table specification to update.
     * @param referenceTableConfigurationName Table comparison name (the reference table configuration name).
     * @param checkType Check type (profiling, monitoring, partitioned).
     * @param checkTimeScale Check time scale.
     */
    public void copyToTableSpec(TableSpec targetTableSpec,
                                String referenceTableConfigurationName,
                                CheckType checkType,
                                CheckTimeScale checkTimeScale) {
        TableComparisonConfigurationSpec comparisonSpec = targetTableSpec.getTableComparisons().get(referenceTableConfigurationName);
        if (comparisonSpec == null) {
            throw new DqoRuntimeException("Table comparison '" + referenceTableConfigurationName + "' was not found in table " + targetTableSpec.getHierarchyId());
        }

        comparisonSpec.setCheckType(checkType);
        comparisonSpec.setTimeScale(checkTimeScale);
        comparisonSpec.setReferenceTableConnectionName(this.getReferenceConnection());
        if (this.getReferenceTable() != null) {
            comparisonSpec.setReferenceTableSchemaName(this.getReferenceTable().getSchemaName());
            comparisonSpec.setReferenceTableName(this.getReferenceTable().getTableName());
        } else {
            comparisonSpec.setReferenceTableSchemaName(null);
            comparisonSpec.setReferenceTableName(null);
        }

        TableComparisonGroupingColumnsPairsListSpec groupingColumnsSpecList = comparisonSpec.getGroupingColumns();
        groupingColumnsSpecList.clear();
        for (TableComparisonGroupingColumnPairModel groupingColumnPairModel : this.groupingColumns) {
            if (groupingColumnsSpecList.size() >= 9) {
                throw new DqoRuntimeException("Too many data grouping columns. DQOps supports up to 9 columns.");
            }
            TableComparisonGroupingColumnsPairSpec groupingColumnsPairSpec = groupingColumnPairModel.createColumnsPairSpec();
            groupingColumnsSpecList.add(groupingColumnsPairSpec);
        }

        AbstractRootChecksContainerSpec tableCheckRootContainer = targetTableSpec.getTableCheckRootContainer(
                checkType, checkTimeScale, true);

        AbstractComparisonCheckCategorySpecMap<?> tableCheckComparisonsMap = tableCheckRootContainer.getComparisons();
        AbstractTableComparisonCheckCategorySpec tableCheckComparisonChecks =
                (AbstractTableComparisonCheckCategorySpec)tableCheckComparisonsMap.getOrAdd(referenceTableConfigurationName);
        if (this.compareRowCount != null) {
            ComparisonCheckRules rowCountMatch = tableCheckComparisonChecks.getCheckSpec(TableCompareCheckType.row_count_match, true);
            this.compareRowCount.copyToComparisonCheckSpec(rowCountMatch);
        } else {
            tableCheckComparisonChecks.removeCheckSpec(TableCompareCheckType.row_count_match);
        }

        if (tableCheckComparisonChecks.supportsColumnComparisonCheck()) {
            if (this.compareColumnCount != null) {
                ComparisonCheckRules columnCountMatch = tableCheckComparisonChecks.getCheckSpec(TableCompareCheckType.column_count_match, true);
                this.compareColumnCount.copyToComparisonCheckSpec(columnCountMatch);
            } else {
                tableCheckComparisonChecks.removeCheckSpec(TableCompareCheckType.column_count_match);
            }
        }

        for (ColumnComparisonModel columnComparisonModel : this.columns) {
            String comparedColumnName = columnComparisonModel.getComparedColumnName();
            if (comparedColumnName == null) {
                throw new DqoRuntimeException("Missing compared column name, it is required");
            }

            ColumnSpec columnSpec = targetTableSpec.getColumns().get(comparedColumnName);
            if (columnSpec == null) {
                throw new DqoRuntimeException("Compared column " + comparedColumnName + " not found on the table " + targetTableSpec.getHierarchyId());
            }

            columnComparisonModel.copyToColumnSpec(columnSpec, referenceTableConfigurationName, checkType, checkTimeScale);
        }
    }

    public static class TableComparisonModelSampleFactory implements SampleValueFactory<TableComparisonModel> {
        @Override
        public TableComparisonModel createSample() {
            TableSpec.TableSpecSampleFactory tableSpecSampleFactory = new TableSpec.TableSpecSampleFactory();
            TableSpec tableSpec1 = tableSpecSampleFactory.createSample();
            TableSpec tableSpec2 = tableSpecSampleFactory.createSample();
            String tableComparisonName = SampleStringsRegistry.getTableComparison();
            TableComparisonConfigurationSpec tableComparisonConfigurationSpec = new TableComparisonConfigurationSpec.TableComparisonConfigurationSpecSampleFactory().createSample();
            tableSpec1.getTableComparisons().put(tableComparisonName, tableComparisonConfigurationSpec);
            return fromTableSpec(
                    tableSpec1,
                    tableSpec2,
                    tableComparisonName,
                    new CheckType.CheckTypeSampleFactory().createSample(),
                    new CheckTimeScale.CheckTimeScaleSampleFactory().createSample(),
                    true);
        }
    }
}

File: distribution/python/dqops/client/models/table_comparison_model.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.check_search_filters import CheckSearchFilters
    from ..models.column_comparison_model import ColumnComparisonModel
    from ..models.compare_thresholds_model import CompareThresholdsModel
    from ..models.delete_stored_data_queue_job_parameters import (
        DeleteStoredDataQueueJobParameters,
    )
    from ..models.physical_table_name import PhysicalTableName
    from ..models.table_comparison_grouping_column_pair_model import (
        TableComparisonGroupingColumnPairModel,
    )


T = TypeVar("T", bound="TableComparisonModel")


@_attrs_define
class TableComparisonModel:
    """Model that contains the all editable information about a table-to-table comparison defined on a compared table.

    Attributes:
        table_comparison_configuration_name (Union[Unset, str]): The name of the table comparison configuration that is
            defined in the 'table_comparisons' node on the table specification.
        compared_connection (Union[Unset, str]): Compared connection name - the connection name to the data source that
            is compared (verified).
        compared_table (Union[Unset, PhysicalTableName]):
        reference_connection (Union[Unset, str]): Reference connection name - the connection name to the data source
            that has the reference data to compare to.
        reference_table (Union[Unset, PhysicalTableName]):
        compared_table_filter (Union[Unset, str]): Optional custom SQL filter expression that is added to the SQL query
            that retrieves the data from the compared table. This expression must be a SQL expression that will be added to
            the WHERE clause when querying the compared table.
        reference_table_filter (Union[Unset, str]): Optional custom SQL filter expression that is added to the SQL query
            that retrieves the data from the reference table (the source of truth). This expression must be a SQL expression
            that will be added to the WHERE clause when querying the reference table.
        grouping_columns (Union[Unset, List['TableComparisonGroupingColumnPairModel']]): List of column pairs from both
            the compared table and the reference table that are used in a GROUP BY clause  for grouping both the compared
            table and the reference table (the source of truth). The columns are used in the next of the table comparison to
            join the results of data groups (row counts, sums of columns) between the compared table and the reference table
            to compare the differences.
        default_compare_thresholds (Union[Unset, CompareThresholdsModel]): Model with the compare threshold levels for
            raising data quality issues at different severity levels when the difference between the compared (tested) table
            and the reference table (the source of truth) exceed given thresholds as a percentage of difference between the
            actual value and the expected value from the reference table.
        compare_row_count (Union[Unset, CompareThresholdsModel]): Model with the compare threshold levels for raising
            data quality issues at different severity levels when the difference between the compared (tested) table and the
            reference table (the source of truth) exceed given thresholds as a percentage of difference between the actual
            value and the expected value from the reference table.
        compare_column_count (Union[Unset, CompareThresholdsModel]): Model with the compare threshold levels for raising
            data quality issues at different severity levels when the difference between the compared (tested) table and the
            reference table (the source of truth) exceed given thresholds as a percentage of difference between the actual
            value and the expected value from the reference table.
        supports_compare_column_count (Union[Unset, bool]): Boolean flag that decides if this comparison type supports
            comparing the column count between tables. Partitioned table comparisons do not support comparing the column
            counts.
        columns (Union[Unset, List['ColumnComparisonModel']]): The list of compared columns, their matching reference
            column and the enabled comparisons.
        compare_table_run_checks_job_template (Union[Unset, CheckSearchFilters]): Target data quality checks filter,
            identifies which checks on which tables and columns should be executed.
        compare_table_clean_data_job_template (Union[Unset, DeleteStoredDataQueueJobParameters]):
        can_edit (Union[Unset, bool]): Boolean flag that decides if the current user can update or delete the table
            comparison.
        can_run_compare_checks (Union[Unset, bool]): Boolean flag that decides if the current user can run comparison
            checks.
        can_delete_data (Union[Unset, bool]): Boolean flag that decides if the current user can delete data (results).
    """

    table_comparison_configuration_name: Union[Unset, str] = UNSET
    compared_connection: Union[Unset, str] = UNSET
    compared_table: Union[Unset, "PhysicalTableName"] = UNSET
    reference_connection: Union[Unset, str] = UNSET
    reference_table: Union[Unset, "PhysicalTableName"] = UNSET
    compared_table_filter: Union[Unset, str] = UNSET
    reference_table_filter: Union[Unset, str] = UNSET
    grouping_columns: Union[Unset, List["TableComparisonGroupingColumnPairModel"]] = (
        UNSET
    )
    default_compare_thresholds: Union[Unset, "CompareThresholdsModel"] = UNSET
    compare_row_count: Union[Unset, "CompareThresholdsModel"] = UNSET
    compare_column_count: Union[Unset, "CompareThresholdsModel"] = UNSET
    supports_compare_column_count: Union[Unset, bool] = UNSET
    columns: Union[Unset, List["ColumnComparisonModel"]] = UNSET
    compare_table_run_checks_job_template: Union[Unset, "CheckSearchFilters"] = UNSET
    compare_table_clean_data_job_template: Union[
        Unset, "DeleteStoredDataQueueJobParameters"
    ] = UNSET
    can_edit: Union[Unset, bool] = UNSET
    can_run_compare_checks: Union[Unset, bool] = UNSET
    can_delete_data: Union[Unset, bool] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        table_comparison_configuration_name = self.table_comparison_configuration_name
        compared_connection = self.compared_connection
        compared_table: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.compared_table, Unset):
            compared_table = self.compared_table.to_dict()

        reference_connection = self.reference_connection
        reference_table: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.reference_table, Unset):
            reference_table = self.reference_table.to_dict()

        compared_table_filter = self.compared_table_filter
        reference_table_filter = self.reference_table_filter
        grouping_columns: Union[Unset, List[Dict[str, Any]]] = UNSET
        if not isinstance(self.grouping_columns, Unset):
            grouping_columns = []
            for grouping_columns_item_data in self.grouping_columns:
                grouping_columns_item = grouping_columns_item_data.to_dict()

                grouping_columns.append(grouping_columns_item)

        default_compare_thresholds: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.default_compare_thresholds, Unset):
            default_compare_thresholds = self.default_compare_thresholds.to_dict()

        compare_row_count: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.compare_row_count, Unset):
            compare_row_count = self.compare_row_count.to_dict()

        compare_column_count: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.compare_column_count, Unset):
            compare_column_count = self.compare_column_count.to_dict()

        supports_compare_column_count = self.supports_compare_column_count
        columns: Union[Unset, List[Dict[str, Any]]] = UNSET
        if not isinstance(self.columns, Unset):
            columns = []
            for columns_item_data in self.columns:
                columns_item = columns_item_data.to_dict()

                columns.append(columns_item)

        compare_table_run_checks_job_template: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.compare_table_run_checks_job_template, Unset):
            compare_table_run_checks_job_template = (
                self.compare_table_run_checks_job_template.to_dict()
            )

        compare_table_clean_data_job_template: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.compare_table_clean_data_job_template, Unset):
            compare_table_clean_data_job_template = (
                self.compare_table_clean_data_job_template.to_dict()
            )

        can_edit = self.can_edit
        can_run_compare_checks = self.can_run_compare_checks
        can_delete_data = self.can_delete_data

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if table_comparison_configuration_name is not UNSET:
            field_dict["table_comparison_configuration_name"] = (
                table_comparison_configuration_name
            )
        if compared_connection is not UNSET:
            field_dict["compared_connection"] = compared_connection
        if compared_table is not UNSET:
            field_dict["compared_table"] = compared_table
        if reference_connection is not UNSET:
            field_dict["reference_connection"] = reference_connection
        if reference_table is not UNSET:
            field_dict["reference_table"] = reference_table
        if compared_table_filter is not UNSET:
            field_dict["compared_table_filter"] = compared_table_filter
        if reference_table_filter is not UNSET:
            field_dict["reference_table_filter"] = reference_table_filter
        if grouping_columns is not UNSET:
            field_dict["grouping_columns"] = grouping_columns
        if default_compare_thresholds is not UNSET:
            field_dict["default_compare_thresholds"] = default_compare_thresholds
        if compare_row_count is not UNSET:
            field_dict["compare_row_count"] = compare_row_count
        if compare_column_count is not UNSET:
            field_dict["compare_column_count"] = compare_column_count
        if supports_compare_column_count is not UNSET:
            field_dict["supports_compare_column_count"] = supports_compare_column_count
        if columns is not UNSET:
            field_dict["columns"] = columns
        if compare_table_run_checks_job_template is not UNSET:
            field_dict["compare_table_run_checks_job_template"] = (
                compare_table_run_checks_job_template
            )
        if compare_table_clean_data_job_template is not UNSET:
            field_dict["compare_table_clean_data_job_template"] = (
                compare_table_clean_data_job_template
            )
        if can_edit is not UNSET:
            field_dict["can_edit"] = can_edit
        if can_run_compare_checks is not UNSET:
            field_dict["can_run_compare_checks"] = can_run_compare_checks
        if can_delete_data is not UNSET:
            field_dict["can_delete_data"] = can_delete_data

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.check_search_filters import CheckSearchFilters
        from ..models.column_comparison_model import ColumnComparisonModel
        from ..models.compare_thresholds_model import CompareThresholdsModel
        from ..models.delete_stored_data_queue_job_parameters import (
            DeleteStoredDataQueueJobParameters,
        )
        from ..models.physical_table_name import PhysicalTableName
        from ..models.table_comparison_grouping_column_pair_model import (
            TableComparisonGroupingColumnPairModel,
        )

        d = src_dict.copy()
        table_comparison_configuration_name = d.pop(
            "table_comparison_configuration_name", UNSET
        )

        compared_connection = d.pop("compared_connection", UNSET)

        _compared_table = d.pop("compared_table", UNSET)
        compared_table: Union[Unset, PhysicalTableName]
        if isinstance(_compared_table, Unset):
            compared_table = UNSET
        else:
            compared_table = PhysicalTableName.from_dict(_compared_table)

        reference_connection = d.pop("reference_connection", UNSET)

        _reference_table = d.pop("reference_table", UNSET)
        reference_table: Union[Unset, PhysicalTableName]
        if isinstance(_reference_table, Unset):
            reference_table = UNSET
        else:
            reference_table = PhysicalTableName.from_dict(_reference_table)

        compared_table_filter = d.pop("compared_table_filter", UNSET)

        reference_table_filter = d.pop("reference_table_filter", UNSET)

        grouping_columns = []
        _grouping_columns = d.pop("grouping_columns", UNSET)
        for grouping_columns_item_data in _grouping_columns or []:
            grouping_columns_item = TableComparisonGroupingColumnPairModel.from_dict(
                grouping_columns_item_data
            )

            grouping_columns.append(grouping_columns_item)

        _default_compare_thresholds = d.pop("default_compare_thresholds", UNSET)
        default_compare_thresholds: Union[Unset, CompareThresholdsModel]
        if isinstance(_default_compare_thresholds, Unset):
            default_compare_thresholds = UNSET
        else:
            default_compare_thresholds = CompareThresholdsModel.from_dict(
                _default_compare_thresholds
            )

        _compare_row_count = d.pop("compare_row_count", UNSET)
        compare_row_count: Union[Unset, CompareThresholdsModel]
        if isinstance(_compare_row_count, Unset):
            compare_row_count = UNSET
        else:
            compare_row_count = CompareThresholdsModel.from_dict(_compare_row_count)

        _compare_column_count = d.pop("compare_column_count", UNSET)
        compare_column_count: Union[Unset, CompareThresholdsModel]
        if isinstance(_compare_column_count, Unset):
            compare_column_count = UNSET
        else:
            compare_column_count = CompareThresholdsModel.from_dict(
                _compare_column_count
            )

        supports_compare_column_count = d.pop("supports_compare_column_count", UNSET)

        columns = []
        _columns = d.pop("columns", UNSET)
        for columns_item_data in _columns or []:
            columns_item = ColumnComparisonModel.from_dict(columns_item_data)

            columns.append(columns_item)

        _compare_table_run_checks_job_template = d.pop(
            "compare_table_run_checks_job_template", UNSET
        )
        compare_table_run_checks_job_template: Union[Unset, CheckSearchFilters]
        if isinstance(_compare_table_run_checks_job_template, Unset):
            compare_table_run_checks_job_template = UNSET
        else:
            compare_table_run_checks_job_template = CheckSearchFilters.from_dict(
                _compare_table_run_checks_job_template
            )

        _compare_table_clean_data_job_template = d.pop(
            "compare_table_clean_data_job_template", UNSET
        )
        compare_table_clean_data_job_template: Union[
            Unset, DeleteStoredDataQueueJobParameters
        ]
        if isinstance(_compare_table_clean_data_job_template, Unset):
            compare_table_clean_data_job_template = UNSET
        else:
            compare_table_clean_data_job_template = (
                DeleteStoredDataQueueJobParameters.from_dict(
                    _compare_table_clean_data_job_template
                )
            )

        can_edit = d.pop("can_edit", UNSET)

        can_run_compare_checks = d.pop("can_run_compare_checks", UNSET)

        can_delete_data = d.pop("can_delete_data", UNSET)

        table_comparison_model = cls(
            table_comparison_configuration_name=table_comparison_configuration_name,
            compared_connection=compared_connection,
            compared_table=compared_table,
            reference_connection=reference_connection,
            reference_table=reference_table,
            compared_table_filter=compared_table_filter,
            reference_table_filter=reference_table_filter,
            grouping_columns=grouping_columns,
            default_compare_thresholds=default_compare_thresholds,
            compare_row_count=compare_row_count,
            compare_column_count=compare_column_count,
            supports_compare_column_count=supports_compare_column_count,
            columns=columns,
            compare_table_run_checks_job_template=compare_table_run_checks_job_template,
            compare_table_clean_data_job_template=compare_table_clean_data_job_template,
            can_edit=can_edit,
            can_run_compare_checks=can_run_compare_checks,
            can_delete_data=can_delete_data,
        )

        table_comparison_model.additional_properties = d
        return table_comparison_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
